clients/src/main/java/org/apache/kafka/common/network/ChannelBuilders.java

```diff
@@ -240,12 +242,19 @@ public static KafkaPrincipalBuilder createPrincipalBuilder(Map<String, ?> config
                                                                KerberosShortNamer kerberosShortNamer,
                                                                SslPrincipalMapper sslPrincipalMapper) {
         Class<?> principalBuilderClass = (Class<?>) configs.get(BrokerSecurityConfigs.PRINCIPAL_BUILDER_CLASS_CONFIG);
-        final KafkaPrincipalBuilder builder;
+        KafkaPrincipalBuilder builder;
 
         if (principalBuilderClass == null || principalBuilderClass == DefaultKafkaPrincipalBuilder.class) {
             builder = new DefaultKafkaPrincipalBuilder(kerberosShortNamer, sslPrincipalMapper);
         } else if (KafkaPrincipalBuilder.class.isAssignableFrom(principalBuilderClass)) {
-            builder = (KafkaPrincipalBuilder) Utils.newInstance(principalBuilderClass);
+            try {
+                Constructor<?> constructor = principalBuilderClass.getConstructor(KerberosShortNamer.class, SslPrincipalMapper.class);
+                builder = (KafkaPrincipalBuilder) constructor.newInstance(kerberosShortNamer, sslPrincipalMapper);
```

Comment:
This looks as if it is solving a problem for a certain Principal type. Not all of the principals have to use KerberosShortNamer, it is related if the credentials are provided by the Kerberos authentication.

Why don't you define a subclass of KerberosPrincipalBuilder instead and provide these details only to that one?

Moreover, you could even define SSLPrincipalBuilder as well.

Reply:
Thanks @plazma-prizma for the feedback,

The above logic will be used whenever user want to config a custom KafkaPrincipalBuilder by extending the existing KafkaPrincipalBuilder.

The goal is to pass already created kerberosShortNamer and sslPrincipalMapper so that User don't miss any functionalities like SSL_PRINCIPAL_MAPPING_RULES_CONFIG when they use custom(own) KafkaPrincipalBuilder.

Reply:
@plazma-prizma, I appreciate your suggestion. To address your concern about the specificity of the solution, I will create Interfaces/subclasses of KerberosPrincipalBuilder and SSLPrincipalBuilder to handle different types of principals and their configurations.

Reply:
@plazma-prizma made the changes as per your inputs, please review when you get a chance, Thanks

Reply:
Thanks for the changes. LGTM.